### PR TITLE
Wave upgrade

### PR DESCRIPTION
**Pull request by ntfrnzn**
Various upgrades, summary to follow

a782223 - add command-line option for wave image; check for image existence (14 hours ago) <Nate Franzen>
a75aeca - update to allow setting current wave-operator image (2 days ago) <Nate Franzen>
fa14ae8 - update kubernetes-ocean-controller (6 days ago) <Nate Franzen>
ab465db - upgrade tide dependency + eksctl binary dependency (12 days ago) <Thorsteinn Thorri Sigurdsson>
e046feb - upgrade tide dep, create and delete tide rbac (3 weeks ago) <Thorsteinn Thorri Sigurdsson>
1c559ae - upgrade tide dependency, delete wave configuration (4 weeks ago) <Thorsteinn Thorri Sigurdsson>
355d0e2 - go mod tidy && go mod vendor (4 weeks ago) <Thorsteinn Thorri Sigurdsson>
8bf6ef0 - remove box and manifests (4 weeks ago) <Thorsteinn Thorri Sigurdsson>
83703d0 - refactor wave/manager (4 weeks ago) <Thorsteinn Thorri Sigurdsson>
d30ae3a - validate cluster context (4 weeks ago) <Thorsteinn Thorri Sigurdsson>
5a4e778 - use tide to delete wave (4 weeks ago) <Thorsteinn Thorri Sigurdsson>
25366a0 - use tide to create wave, upgrade eksctl binary dependency (4 weeks ago) <Thorsteinn Thorri Sigurdsson>
4ef50a6 - include version in third-party dependency (5 weeks ago) <Nate Franzen>
a79fea0 - sparkoperator -> spark-operator (5 weeks ago) <Thorsteinn Thorri Sigurdsson>
b00a41b - bump wave operator chart version (0.1.8) (5 weeks ago) <Thorsteinn Thorri Sigurdsson>
75fc42b - use spot-hosted spark-operator chart (6 weeks ago) <Nate Franzen>

Refactoring
- The management code for installing and deleting wave has been moved
into the wave-operator repository, since that's a more natural grouping.
spotctl now imports a package called "tide"
- spotctl can now set the initial image to use for the wave-operator,
and there's a optional command-line flag to specify it. spotctl imports
github.com/google/go-containerregistry so that it can [check whether the
image exist](https://github.com/spotinst/spotctl/blob/wave-upgrade/internal/cmd/wave/create.go#L145)s before starting installation (it would be painful to go through
the entire 20-minute EKS process before finding out that the image was
incorrect).

Functionality
- spotctl explicitly installs latest ocean-controller. Currently, a particular
version of the ocean controller is intalled by the version of eksctl. It's
a slow process to upgrade this reference -- much easier to apply the manifests
directly from the S3 bucket. To do this, spotctl now has an [internal/kubernetes](https://github.com/spotinst/spotctl/blob/wave-upgrade/internal/kubernetes/apply.go)
package which downloads the manifests and uses a kubernetes dynamic client to 
apply them.
- describe output is improved, writing the component properties in a table

### CODE REVIEW

**Comment by ntfrnzn** [comment]
By the way, the server-side apply code is pretty much copied from https://ymmt2005.hatenablog.com/entry/2020/04/14/An_example_of_using_dynamic_client_of_k8s.io/client-go which -- for me -- was really helpful in clarifying some k8s-api concepts.

File: internal/cmd/wave/create.go [hunk 1]
```diff
@@ -531,3 +622,40 @@ func allNonDeletedStackStatuses() []string {
...
 }
+
+func updateOcean(ctx context.Context, logger logr.Logger) error {
+
+	conf, err := config.GetConfig()
+	if err != nil {
+		return fmt.Errorf("cannot get cluster configuration, %w", err)
+	}
+
+	const oceanURL = "https://s3.amazonaws.com/spotinst-public/integrations/kubernetes/cluster-controller/spotinst-kubernetes-cluster-controller-ga.yaml"
```
Comment (ntfrnzn) [inline internal/cmd/wave/create.go:633]: really questionable to put this URL in the code but we can work forward from here I think.

File: internal/dep/defaults.go [hunk 2]
```diff
@@ -22,7 +22,7 @@ var (
 	// See: https://github.com/weaveworks/eksctl.
 	DependencyEksctl Dependency = &dependency{
 		name:    "eksctl",
-		version: "0.31.0-fc20af78",
+		version: "0.36.2-f5c273f8",
```
Comment (ntfrnzn) [inline internal/dep/defaults.go:25]: 😢 to put the versions in code here. TODO refactoring somehow.

File: internal/cmd/wave/create.go [hunk 3]
```diff
@@ -531,3 +622,40 @@ func allNonDeletedStackStatuses() []string {
...
+		return fmt.Errorf("error fetching ocean manifests, %w", err)
+	}
+	data, err := ioutil.ReadAll(res.Body)
+	defer res.Body.Close()
+	if err != nil {
+		return fmt.Errorf("error reading ocean manifests, %w", err)
+	}
+
+	delim := regexp.MustCompile("(?m)^---$")
+	objects := delim.Split(string(data), -1)
```
Comment (ntfrnzn) [inline internal/cmd/wave/create.go:646]: not sure why the yaml libraries don't split on "---" according to the spec, but that's where we are

**Review by liranp — APPROVED** [review]

**Pull request merged**